Generic export module

yecli opened the pull request:
Generic export module added

yecli committed:
#334: Initial projects structure

yecli committed:
#334: Added some core models;

yecli committed:
#334: Added independent ExportModule solution;
Removed redundant dependencies;

akak1977 committed:
Export/335 implement exported type metadata provider (#338)

#335: 1st attempt to implement

yecli committed:
#340: Added and registered service that implements IKnownExportTypes-Registrar/-Provider; (#341)

#340: Added and registered service that implements IKnownExportTypes-Registrar/-Provider;

akak1977 committed:
#335: implement exportedtypemetadata provider (#346)

Implemented `ExportedTypeMetadata.GetFromType<T>()`

AliveMen committed:
JsonExportProvider moq (#351)

Added JsonExportProvider stub

yecli committed:
Export 342 data exporter implementation (#348)

Added DataExporter

yecli committed:
#349: Implemented CreateProvider - it could create Json and Csv providers (#352)

Implemented CreateProvider

akak1977 committed:
Export/339 pricing module export integration (#353)

- PriceExportPagedDataSource implemented;
- Added workflow test;

akak1977 committed:
#335: ExportedTypeMetadata improvements;

- Stops recursive walking around Entity-derived properties;
- Search order for Entity-derived properties changed;

akak1977 committed:
#345: First implementation of CSVExportProvider (#362)

First CSVExportProvider version implemented

AliveMen committed:
#336: JsonExportProvider (#359)

#336: Implemented JsonExportProvider first version

AliveMen committed:
354 add Pricelist and PricelistAssigments infrastructure (#365)

#354: Add Pricelist's and PricelistAssigment's export infrastructure

yecli committed:
Export 347 export module api (#366)

#347: Imlemented API;
- Reworked ExportProviderFactory;

yecli committed:
Export 347 export api part 2 (#371)

* #347: API + fixed number of errors:
- Error with paging in BaseExportDataSource;
- Moved ExportDataQuery to Platform.Core to support DataQuery polymorhic deserialization in Hangfire;
- Removed redundant ExportMetadata properties from Kson serialization;
- Properly registered KnownExportTypeService;

AliveMen committed:
#354: Update Price Json Export test

Updated PriceJson Export test

AliveMen committed:
#370: Export/fix exported values is null (#373)

add default valued for IncludedProperties
filter PropertiesInfo without changing original metadata object

yecli committed:
#345: Export csv provider (#380)

* #345: Added configurable CsvExportProvider based on CsvHelper.CsvWriter, which automatically filter properties based on ExportedTypeMetadata properties;
It skips IEnumerable<Entity> properties as there is no good generic mechanism to write them to CSV. Entity nested properties are exported; 
Added tests;

AliveMen committed:
dev Export buttons on Prices blade (#377)

Added export blade to ExportModule;
Implemented prices export;

AliveMen committed:
use priceListIds instead priceListId (#390)

use priceListIds instead priceListId

yecli committed:
Removed exception in WriteMetadata in CsvExporter;

AliveMen committed:
Export/dev export pricelists and assigments (#394)

export pricelists and assigments

AliveMen committed:
fill Pricelist.Prices (#395)

fill Pricelist.Prices

yecli committed:
Export 387 different object types in one export file (#396)

* #387: Supported different objects types writing in CSV and Json;
Added ExportedFileExtension to IExportProviderConfiguration;
Fixes for resulting array in JsonExportProviders tests;

yecli committed:
Export 387 different object types in one export file (#397)

- Used ExportName as column name in CsvProvider, updated tests;
-Removed different type handling possibility by CsvProvider as mapping becomes too complex;
- Added test for CsvProvider for same type properies mapping;
- Minor refactoring for ExportTypeMetadata;

yecli committed:
#386: Infrastructure changes (#408)

Moved stream from ExportProviderFactory, added TextWriter to IExportProvider.Write;
Extended and fixed some export tests;
Now files are exported with proper extensions;

yecli committed:
#386: Removed ExportDataQueryfrom platform, extended Hangfire serializer options in Export module; (#409)

Used own Export module ProgressInfo and notification instead of the Platform ones.

AliveMen committed:
#405: Export/dev export type selector (#412)

Implemented blade with Exported type selector

yecli committed:
Export 398 architecture changes (#418)

#398: Updated infrastructure architecture, added "Pricelist Full Data" to Pricing export possibilities;

yecli committed:
Changed lifetime of DataExporter (no need in transient);

yecli committed:
#404: export main blade (#427)

Main export blade added.

yecli committed:
#404: Export main blade style adjustments;

yecli committed:
Rebased on the latest dev, fixed erros

akak1977 committed:
Export/401 security part (#431)

Export module permissions: access, download added.
Security check added to BaseExportPagedDataSource.
ExportDataSources initialization in PricingModule changed according to BaseExportPagedDataSource changes.

akak1977 committed:
407. Export column selector completed (#438)

#407. Export column selector completed

yecli committed:
Export 406 generic object viewer (#446)

Implemented generic object viewer for ViewableEntity using IPagedDataSource.GetData() method;
Implemented IPagedDataSource.GetData() for Pricing module data sources;
Changed workflow - provider should be selected before data and columns;

AliveMen committed:
#425: Used cloned objects in export (#447)

Data sources provide ICloneable now, we are changing clones in DataExporter

yecli committed:
#407: Excluded nested metadata for Price and PricelistAssignment;

akak1977 committed:
#437 TabularMetadata (#448)

#437: TabularMetadata added. Now we can select columns from a range dependent on provider type. Also we need to select provider before all other export options.

AliveMen committed:
#453: Generic export form was attached to pricing module

Generic export form was attached to pricing module

akak1977 committed:
#450: Export Catalog implemented (#454)

Export Catalog implemented

yecli committed:
#444: custom generic filter (#456)

Added filter customization mechanism;
Added filters for Pricing module (Pricelists, Products);

yecli committed:
#458: Added Parent and Type to ViewableEntity; (#462)

Added Parent and Type to ViewableEntity;
Fixed ExportPriceCSV test;

AliveMen committed:
#445: Export/fix export prices (#464)

Fixed "selected: 0" and empty types blade

yecli committed:
#466: Generic viewer buttons and scroll changes; (#468)

Removed generic viewer buttons;
Replaced pagination with infinite scroll;
Removed "Manage" action and context menu from viewer template;

yecli committed:
#466: Fixed refresh to reset selection and keyword filtration;

yecli committed:
#466: Preserving selected items when close/open viewer;

akak1977 committed:
VP-227: Column selector rework completed (#471)

Reworked column selector using mutliselect combos

yecli committed:
vp-225: generic viewer extension (#473)

* vp-225:
Reworked ViewableEntities machanics, adapted all Pricing objects to it;
Removed sorting from grid as some entities could not be sorted by some fields when querying data, because fields are filled after we get data (e.g. Price.Name, Price.Pricelist, Price.Product etc.);

yecli committed:
Reverted Export changes for Catalog module for now.

yecli committed:
Changed Task.Result to GetAwaiter().GetResult() for clear exceptions;

AliveMen committed:
vp-223: rework export setting blade (#475)

Updated UI for export-settings blade
type selection first;
receive data total count;
save selected provider to local storage;

yecli committed:
VP-234: Fixed filters clearance

AliveMen committed:
vp-223: fixed selected item count mistake (#478)

akak1977 committed:
Export/vp 228 export type names localization (#479)

* Added localization for export provider names, export type names and its groups.
* allColumnsOfType removed from exportDataRequest to avoid redundant data in export controller request

yecli committed:
Fixed error when opening export type selector without selected provider;

AliveMen committed:
Export/vp229 Add advanced filters to PricelistAssignments  (#481)

Add advanced filter to PricelistAssignments

yecli committed:
VP-230: Added generic viewer item viewers;

Added generic viewer item viewers;

akak1977 committed:
VP-231: CSV column filtering implemented. (#483)

CSV column filtering implemented;
Resolved bugs with provider selector on an export-settings blade;

yecli committed:
VP-228: Fixed types names in filter metadields registration;

AliveMen committed:
Fixed selected items count (#484)

Fixed selected items count

yecli committed:
VP-243: Added Apply button for filter details; (#485)

Added Apply button for filter details;
Fixed filters loading/saving;
Removed default "UnnamedFilter" filter name;
Filter creationform is shown on clicking on pencil (edit) icon now;

akak1977 committed:
VP-248: Descriptions for export types and groups implemented. (#489)

* Descriptions for export types and groups implemented

* Beating bug with irrelevant column sets while changing export provider tabular->nontabular and vice-versa

* Review-based changes.
Simplify export-settings blade logic.

yecli committed:
VP-247: Changed selection logic - now in case of now selected rows 0 rows would be selected, all rows would be selected based on all selected checbox. (#491)

yecli committed:
VP-249: Made export screen non closable;

akak1977 committed:
Redundant types filtering removed from export-type-selector blade

tatarincev committed:
fix bug with a exception was throw  on  a license download

yecli committed:
Fixed PricingCSV tests;

yecli committed:
Ignore module build output;

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Core/Model/BaseExportPagedDataSource.cs:
```diff
@@ -0,0 +1,132 @@
+
+    public abstract class BaseExportPagedDataSource : IPagedDataSource
+    {
+        protected class FetchResult
```
Why do we need this embedded type? IPagedDataSource already has GetTotalCount() method for this purposes and total records count must be immutable when a iterator is initialized and used for data fetching.

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Core/Model/BaseExportPagedDataSource.cs:
```diff
@@ -0,0 +1,132 @@
+        /// Permissions should be attached to *ExportPagedDataSource class using <see cref="AuthorizeAttribute"/>
+        /// </summary>
+        /// <returns></returns>
+        public virtual async Task Authorize()
```
Include Authorization logic in this type will leads to reduce cohesion of this type and as result a wide responsibility. I'll suggest to move out any authorization logic from here

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Core/Model/BaseExportPagedDataSource.cs:
```diff
@@ -0,0 +1,132 @@
+
+        protected abstract FetchResult FetchData(SearchCriteriaBase searchCriteria);
+
+        protected abstract ViewableEntity ToViewableEntity(object obj);
```
This methods also looks a redundant and doesn't fit the primary intention of this type.  Why we can't perform a logic of mapping exported types into `ViewableEntity`  in other place or  more why doe  we need this type `ViewableEntity` at all, Maybe better will be use a special interface `IExportViewable` and inherit it for all exported data types?

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Core/Model/ExportedTypeDefinition.cs:
```diff
@@ -0,0 +1,51 @@
+        [JsonIgnore]
+        public Func<ExportDataQuery, IPagedDataSource> ExportedDataSourceFactory { get; protected set; }
+
+        public ExportedTypeDefinition WithDataSourceFactory(Func<ExportDataQuery, IPagedDataSource> factory)
```
Better use for this a special builder which a wrapped an object and provides a fluent syntax for it configuration  through own or extension methods.
https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreBuilder.cs

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Core/Model/ExportedTypeMetadata.cs:
```diff
@@ -0,0 +1,118 @@
+        }
+
+
+        private ExportTypePropertyInfoEx[] GetFromType(Type type, string groupName, string baseMemberName, List<MemberInfo> passedNodes, bool extractNestedProperties)
```
Better move out this logic from here to extension method and think about reduce of complicity this logic.

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Core/Services/IKnownExportTypesRegistrar.cs:
```diff
@@ -0,0 +1,10 @@
+    public interface IKnownExportTypesRegistrar
+    {
+        ExportedTypeDefinition[] GetRegisteredTypes();
+        ExportedTypeDefinition RegisterType(string exportedTypeName, string group, string exportQueryType);
```
need to remove all formal parameters from this method and pass `ExportedTypeDefinition`  directly in ctor. If we can't make this method as generic and without parameters.

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Core/Model/ExportDataQuery.cs:
```diff
@@ -0,0 +1,47 @@
+        public int? Skip { get; set; }
+        public int? Take { get; set; }
+
+        public abstract SearchCriteriaBase CreateSearchCriteria();
```
We don't need this virtual method here, query will passed directly in a paged data source and will criteria have to constructed there by using `AbstractTypeFactory`  in order to preserve the possibility of using overridden types of criteria.

tatarincev reviewed (commented):
Please add comments for each domain types with brief description of their responsibility

yecli committed:
VP-247: Fixed bug with improper select all behavior after 0 items was selected previously.

yecli committed:
Fixed bug when we export all data in case 0 selected by disabling Start export in case of 0 selected.

yecli committed:
VP-160: Export button is dusabled while current export is in progress.

AliveMen committed:
VP-229: add advanced filters to pricing module viewers (#495)

add catalog filter for assignments;
add currencies filter to Priselists;
fix problem with opening export from the price module blade;

akak1977 committed:
Export/vp 250 replace get from type with reflection utility (#496)

GetFromType and related moved to different helper in ExportModule.Data

AliveMen committed:
VP-229: fix request sequence (#497)

yecli committed:
VP-253: Reworked ExportedTypeDefinition registration, left non generic way in case we want to specify different group for some type (e.g. Pricing object in Catalog group); (#499)

AliveMen committed:
vp-252 Authorization rework. (#498)

Authorization rework.

AliveMen committed:
use serviceCollection.Configure for MvcJsonOptions (#502)

yecli committed:
Added Platform.Secutity project to Pricing solution to fix its build

akak1977 committed:
Export/vp 251 i paged data source rework (#500)

GetFromType and related moved to different helper in ExportModule.Data
Deep refactoring around export types, backend for universal export viewer, pricing export types, etc.

yecli committed:
VP-251: datasources rework (#505)

Inherited Exportable objects from sepcific data objects we want to export, e.g. ExportedPrice : Price, IExportable;
Added abstract SingleTypeExportDataSource to Export.Data as base class for single type datasources;
Registered Datasources and factories as Transient;
Cloned reference properties in IExportable implementations;

yecli committed:
Merge branch 'dev' into export/dev

yecli committed:
Properly fixed auto-merge conflicts;

yecli committed:
VP-251: Removed metadata ExportablePricelistFullData registration for now, as it will be inplemented properly in VP-255;

akak1977 committed:
Export/vp 254 xml comments (#508)

ExportedTypeColumnInfo =>ExportedTypePropertyInfo
ExportedTypePropertyInfo.Name => FullName
ExportedTypePropertyInfo.ExportName => DisplayName

yecli committed:
#511: Returned auto-removed references;

yecli committed:
VP-251: Fixed pagination problem;

yecli committed:
Merge remote-tracking branch 'origin/dev' into export/dev

yecli committed:
Merge branch 'dev' into export/dev

akak1977 committed:
vp 254 xml comments (#513)

* Comments in progress

* Columns -> Properties

* Comment for ExportDataQuery.ExportTypeName

* Columns->Properties 2

* Column -> Property 3

* Columns->Properties 4

* Comments add/rework

* Healed the bug with dropping included properties after use select data blade

* Review-based changes

* Review-based changes 2

yecli committed:
Revert "vp 254 xml comments (#513)" (#514)

This reverts commit a0d56e164b0a8031d60de0d2aaa6ac01f20a4995.

yecli committed:
Revert "Revert "vp 254 xml comments (#513)" (#514)" (#515)

This reverts commit fa9c49645aa71c256bad6f7f94dbf914a943967c.

akak1977 committed:
VP-288 Export from pricing blades now works (#516)

yecli committed:
VP-254: Reverted redundant renaming;

akak1977 committed:
VP-254: xml comments (#517)

 VP-254: xml comments
Used ExportableSearchResult wehre possible;

yecli committed:
Removed redundant Test.MockHelpers project

akak1977 committed:
VP-277: Metadata retrieving reworked (#519)

Metadata retrieving reworked

yecli committed:
Refactored ExportedTypeDefinitionBuilder, added comments;

yecli committed:
More ExportedTypeDefinitionBuilder refactoring;

yecli committed:
Fixed product selector title in pricing filter;

yecli committed:
Added XML-comments to SingleTypeExportPagedDataSource;

yecli committed:
Merge branch 'dev' into export/dev

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Core/Model/IExportable.cs:
```diff
@@ -0,0 +1,17 @@
+    /// Interface to implement exportаble entities.
+    /// Common (default) set of properties for universal viewing.
+    /// </summary>
+    public interface IExportable : ICloneable
```
`IExportable` interface shouldn't contains fields are intended for presentation in UI it should be a separate interface `IExportViewable`  or some like that.
`IExportable` must be or empty or be derived from `IEntity` and `ICloneable` interfaces. And all export services must operate `IExportable` interfaces instead of object

tatarincev reviewed (commented):
Could we move csv and json export providers into separate projects within this module, it will be a good example of how to implement custom provider and leads a good architecture and responsibility segregation?

tatarincev committed:
Merge branch 'export/dev' of https://github.com/VirtoCommerce/vc-platform-core into export/dev

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Data/Services/CsvExportProvider.cs:
```diff
@@ -0,0 +1,171 @@
+    /// Does not map <see cref="IEnumerable<Entity>"/> as these are not representable in CSV structure in suitable manner.
+    /// </summary>
+    /// <typeparam name="T">Mapped type.</typeparam>
+    public class MetadataFilteredMap<T> : ClassMap<T>
```
move this class map out from this file to a new  file

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Data/Services/JsonExportProvider.cs:
```diff
@@ -0,0 +1,128 @@
+            {
+                var jObject = (JObject)jToken;
+
+                jObject.AddFirst(new JProperty("$discriminator", value.GetType().FullName));
```
We should avoid using the full type name in serialization, because there is an extension model for domain types in our system, and there can be overridden types in the target system that are constructed based on the name of the base type trough AbstractFactory<>. So use  value.GetType().Name instead

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Data/Services/JsonExportProvider.cs:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+
+    public class ObjectDiscriminatorJsonConverter : JsonConverter
```
Move to a new file

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Data/Services/JsonExportProvider.cs:
```diff
@@ -0,0 +1,128 @@
+                jsonSettings = jsonProviderConfiguration.Settings;
+            }
+            else
+            {
```
Better move this  `else statement`  default settings initialization into ExportProviderConfiguration (do this initialization with default values on it side)

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Core/Services/IExportProviderFactory.cs:
```diff
@@ -0,0 +1,12 @@
+    /// </summary>
+    public interface IExportProviderFactory
+    {
+        IExportProvider CreateProvider(string name, IExportProviderConfiguration config, ExportedTypePropertyInfo[] includedProperties);
```
Let's remove all of these formal parameters and use  only one with `ExportDataRequest` type

tatarincev commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Data/Services/SingleTypeExportPagedDataSource.cs:
```diff
@@ -0,0 +1,94 @@
+    /// </summary>
+    /// <typeparam name="TDataQuery">Specific <see cref="ExportDataQuery"/> type that is used to query this data source.</typeparam>
+    /// <typeparam name="TSearchCriteria">Specific <see cref="SearchCriteriaBase"/> type that is used by this data source to query data.</typeparam>
+    public abstract class SingleTypeExportPagedDataSource<TDataQuery, TSearchCriteria> : IPagedDataSource
```
I would like to rename this method to abstract  ExportDataPageSourceBase  with protected ctor that received 
 `TDataQuery` as single formal  parameter and encapsulate it.

Method `FillSearchCriteria` must be removed and left only one `BuildSearchCriteria` for a criteria construction.
this class should looks like on this image 
![image](https://user-images.githubusercontent.com/7566324/63163303-58f0a680-c025-11e9-80aa-9aaebfac7897.png)

yecli committed:
Removed redundant comment;

yecli committed:
VP-296: IExportable changes (#527)

No props in IExportable;
Added IExportViewable;
Providers work with IExportable instead of object;
Added ITabularConvertible;
Removed TabularDataConverters, used ITabularConvertible by exported entities instead;
Added reference names to TabularPrice and TabularPricelistAssignment;
Reorganizaed IExportProviderFactory with single data request param;
Provider config default values filling to specific config classes constructors;

akak1977 committed:
VP-291: wrong number of selected objects (#526)

Search criterias now consistently passing between blades
Generic viewer: take-skip hell removed

AliveMen committed:
VP-298: single type export paged data source rework (#528)

Rename SingleTypeExportPagedDataSource to ExportPagedDataSource;
Exclude FillSearchCriteria in favor of overriding BuildSearchCriteria with calling base;
Updated Data source factories;

AliveMen committed:
VP-301: IPagedDataSource enumeration rework (#529)

Changed enumeration methods from
IEnumerable<IExportable> FetchNextPage();
to
bool Fetch();
IEnumerable<IExportable> Items;

akak1977 committed:
VP-291: wrong number of selected objects (#530)

* Search criterias now consistently passing between blades
* Generic viewer: take-skip hell removed
* Selecting nothing without changing search criteria

yecli committed:
VP-291: Fixed server pagination cycling;
Clear pagination on export on the client;

akak1977 committed:
Export/vp 297 move providers to separate projects (#532)

Providers moved to separate projects;
Targeted Marketing project to latest minor version to compile in VS 2017;

yecli committed:
Merge branch 'dev' into export/dev

yecli committed:
Formatting changes;

yecli committed:
Changes for extended classes export support:
- ExportedTypeDefinitionBuilder could be created with existing ExportedTypeDefinition;
- Made ToModel() virtual for pricing exportable entities;

yecli committed:
VP-299: Added "Export.General.FileNameTemplate" setting to manage export file name template. (#535)

Added "Export.General.FileNameTemplate" setting to manage export file name template.

yecli committed:
VP-297: Moved specific provider configurations to providers;
Removed redundant packages from Export.Data and added to providers;
Excluded providers config from Json for now, as default configs took 86MB of traffic! Need to rework config with only specific settings then;

yecli committed:
VP-316: Filters save fix.

yecli committed:
VP-353: Eliminated ccycling by proper paging support for query with ObjectIds.

yecli committed:
VP-224: Apply is disabled for saved filters;

yecli committed:
VP-244: Apply is disabled when editing saved filter;

akak1977 committed:
Product-selector now correctly shows number of products (#541)

Product-selector now correctly shows number of products

akak1977 committed:
VP-355: Export download OR-combined policy check implemented (#538)

OR-combined policy check implemented and applied to export download.

yecli committed:
Merge branch 'dev' into export/dev

yecli committed:
Correctly merged Pricing.Web Module.cs

AliveMen committed:
VP-366: export products (#543)

Added ability to export products to csv and json

AliveMen committed:
VP-376: Cannot close export progress blade. (#544)

Remove isClosingDisabled: true;

yecli committed:
VP-362: Created specific IPagedDataSourceFactory classes for simpler datasource factories injection;

yecli committed:
VP-362: ExportDataQuery implementation registrations to ExportedTypeDefinitionBuilder;

AliveMen committed:
Export/VP-367 product export ui (#551)

updated product search criteria
used scope based authorization
added product's filters
added "reset" button to category-selector blade.
added search in children categories

akak1977 committed:
VP-364: catalog category export (#549)

VP-364: Added full catalog data export;
Created ComplexExportPagedDataSource base class for complex data sources in Export module;

AliveMen committed:
VP-369: Product export in catalog category/product blade; (#554)

VP-369: Integrated Product export into catalog category/product blade;

yecli committed:
VP-364: Removed redundant Catalog.Tests project;

yecli committed:
VP-364: Removed redundant Catalog.Test.csproj file;

yecli committed:
VP-364: Full data source refactoring (#559)

Moved export models from Data to Core project.
Added extensions instead of ComplexExportPagedDataSource;
Refactored MultipleDataSources tests;
Full data source major refactoring;
Added int? Skip and Take to IPagedDataSource for proper pagination and querying logic;
Fixed typo in "IPropertyDictionaryItemSearchService" name;

yecli committed:
VP-367: Added viewer for product;

yecli committed:
VP-364: Fixed filters and selectors caused by changed type names;
Set searchInChildren = true for export from "Categories & Producs" blade;

tatarincev commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -0,0 +1,85 @@
+            _productDataSourceFactory = productDataSourceFactory;
+            _dataQuery = dataQuery;
+
+            _dataSources = CreateDataSources();
```
Remove call of virtual methods from ctor

tatarincev committed:
Minor architectural refactoring to improve catalog export extensibility

yecli committed:
VP-399: Refactoring - used single factory instead of one for all data sources for brievity;

yecli committed:
VP-399: Fixed problem with improper registration of CatalogFullExportDataQuery and its forner descentant ProductExportDataQuery;

yecli committed:
VP-399: Handling binary data in product full data source fix;

yecli committed:
VP-399: IPagedDataSourceCollectionExtensions semantic fix
